- Convert SumF functions for Array, Span, and List into a generic function

Smurf-IV opened the pull request:
- Convert SumF function into a generic function
- Add new managed types (sbyte, byte, short, ushort, int, uint, long, ulong)
- Add tests for some of the new types
- Add new Benchmark tests

Commit:
- Convert SumF function into a generic function
- Add new managed types (sbyte, byte, short, ushort, int, uint, long, ulong)
- Add tests for some of the new types
- Add new Benchmark tests

Smurf-IV commented:
This fixes #18 
[Bug] using SumS on a ushort[] throws System.InvalidCastException #18

Commit:
SumF for an Array
- Make sure publics have Intellisense
- Add Nullable types
- Perform template on source selectors

Commit:
- Split the Sum types out
- Apply template to Spans<t>
- Apply Template to List<T>

Commit:
- Fix float selector special case
- Add List benchmarks

Commit:
[Enhancement] Can these API's also work on IList<T> #15
- Add SumF API's to redirect IList's to their base types

Commit:
- Add Nullable types for List and IList

jackmott commented:
Hey Simon, wanted to say thanks for all the experiments you are doing here.  I'll need to some time to review/test some of this stuff out, and I'm pretty busy with family/work all the time so it might take a while. I'll try to take a deep look in a few days and let you know what I think.

Commit:
- Use "IList fast redirect" to implement all average Base and Nullable types

Smurf-IV commented:
> wanted to say thanks for all the experiments you are doing here. 

No worries, Now that I have the templates in place, I can burn through the other types pretty quickly

Commit:
- Complete base types applicable to Span<T>
- Add inlining hint to AverageIList

jackmott commented:
I would hold off on doing too much work until I've had some time to review and test.  I don't want you to write a million lines of code and then I end up not using them!

Commit:
Fix #21: [Bug] `Span SumF` is slower than a for loop over a span
- Fix code to make it appear "As fast" Via the use of the "Coming out faster" acum = accum + nextVal method
- Add ReadOnlySpan to `SumF` api's

Commit:
[Enhancement] Can these API's also work on IList<T> #15
- Done for Aggregate
- Add All Span types
- Add ReadOnlySpan
- Reduce code in SumList

Commit:
[Enhancement] Can these API's also work on IList<T> #15
- Fix copy paste error
- Add better intellisense text

Commit:
Investigate #22: [Question] Span_FirstF and List_First are slower than just calling foreach -> Why?

Commit:
Investigate #22: [Question] Span_FirstF and List_FirstF are slower than just calling foreach -> Why?

Commit:
Investigate #15: [Enhancement] Can these API's also work on IReadOnlyList<T>
- Change over to IReadOnlyList

Commit:
More for #15: [Enhancement] Can these API's also work on IReadOnlyList<T>
- Implement LastF's the same way as FirstF's

Commit:
Initial commit for #24: [Enhancement] DefaultIfEmptyF should be implemented
- Add Start File

Commit:
MaxF's for  #15 : [Enhancement] Can these API's also work on IReadOnlyList<T>
- Use types instead of vars
- Breakout Max Benchmarks
- Breakout MaxF API's for different types into seperate files

Commit:
MinF's for #15 :  [Enhancement] Can these API's also work on IReadOnlyList<T>
- Split MinF's out into seperate files
- Change to NameOf for Selector parameter

Commit:
WhereSumF's for  #15 : [Enhancement] Can these API's also work on IReadOnlyList<T>
- Replacement to use nameof for source
- Split more benchmarks out into seperate relevant files

Smurf-IV commented:
> 
> 
> Hey Simon, wanted to say thanks for all the experiments you are doing here. I'll need to some time to review/test some of this stuff out, and I'm pretty busy with family/work all the time so it might take a while. I'll try to take a deep look in a few days and let you know what I think.

Just found a bog "No-no" in some of the "Optimisations" I've done
When using the List's I have done the following

            int sourceCount = source.Count;
            T2 a = p.Zero();
            checked
            {
                for (int index = 0; index < sourceCount; index++)
                {
                    a = p.Add(a, source[index]);
                }
            }

The big "No-No" is on the ` int sourceCount = source.Count;` as the Items in the List could be changed externally "Darn threading !". One way would be to lock the list, another is to put the count Back into the for loop,
I'll go with the latter, I think this may also need to be done for the IReadOnlyList but I'm not sure.

Smurf-IV commented:
// Note: Lists can have items added and removed whilst these API's are in use
        // The IReadOnlyList<T> represents a list in which the _number_ and _order_ of list elements is read-only.
        //

Commit:
- Note: Lists can have items added and removed whilst these API's are in use
- The IReadOnlyList<T> represents a list in which the _number_ and _order_ of list elements is read-only.

jackmott commented:
`int sourceCount = source.Count;`  is a non optimization on List anyway. Did you mean IList?

Smurf-IV commented:
> 
> 
> `int sourceCount = source.Count;` is a non optimization on List anyway. Did you mean IList?

No on the `List<T>` it did show "Some" improvement....

Commit:
Invesigate #25: [Enhancement] Please also target .net4.8 and Benchmark
- Add Ratio Comparisons to Benchmarks
- Add .Net4.8 Output directories
- Explicit call types for Average

Commit:
Investigate #25: [Enhancement] Please also target .net4.8 and Benchmark
- Attempt to sort why Appveyor is unhappy.
- Sort out the ratio detection code when all are put back in.

Smurf-IV commented:
No idea why AppVeyor is complaining, It runs on my machine !!

Commit:
Investigate #21: [Bug] Span SumF is slower than a for loop over a span
Create #22: [Question] Span_FirstF and List_FirstF are slower than just calling foreach -> Why?
Span this #24: [Enhancement] DefaultIfEmptyF should be implemented
Add Span #25: [Enhancement] Please also target .net4.8 and Benchmark

Commit:
- Attempt to fic AppVeyor

Commit:
- Another Attempt to satisfy Appveyor

Commit:
- Appveyor again - Reset the bindingRedirects

Commit:
- Look Apveyor Use the correct system.memry version will ya !

Commit:
- (Maybe) Seems like the App.config is being ignored, so be explicit in the csproj file...

Commit:
- Fix weird "Sometimes failing" test that show aggressive inlining doing weird stuff.

Commit:
- test the unit Test Again !

ndrwrbgs commented on LinqFaster/Select.cs:
```diff
@@ -245,12 +143,12 @@ public static void SelectInPlaceF<T>(this List<T> source, Func<T, int, T> select
             if (source == null)
             {
-                throw Error.ArgumentNull("source");
+                throw Error.ArgumentNull(nameof(source));
```
isolating the changes would be a major help to reviewers. These, while potentially desirable, provide a lot of noise as they aren't "Convert SumF into a generic function"

ndrwrbgs commented on Benchmarks/App.config:
```diff
@@ -1,90 +1,30 @@
+    <startup>
         <supportedRuntime version="v4.0" sku=".NETFramework,Version=v4.6.1" />
     </startup>
-  <runtime>
```
Avoiding these kind of whitespace changes can help focus the review and speed up the review. In addition, it helps reviewers to stay focused so they can more likely catch issues

ndrwrbgs commented on LinqFaster/Repeat.cs:
```diff
@@ -34,7 +34,7 @@ public static T[] RepeatArrayF<T>(T element, int count)
         public static List<T> RepeatListF<T>(T element, int count)
         {
-            var result = new List<T>(count);
+            List<T> result = new List<T>(count);
```
more 'style' changes that aren't necessary

ndrwrbgs commented on LinqFaster/Aggregate.cs:
```diff
@@ -9,149 +13,76 @@ namespace JM.LinqFaster
+        /// <param name="source">An IReadOnlyList to aggregate over.</param>
         /// <param name="func">An accumulator function to be invoked on each element</param>
         /// <returns>The final accumulator value</returns>
-        public static TSource AggregateF<TSource>(this TSource[] source, Func<TSource, TSource, TSource> func)
```
Converting T[] to IReadOnlyList<T> makes it have to call through the interfaces for everything, which involves a vtable lookup. If this is a change (and not a reformatting artifact of moving things etc), then this would almost surely regress performance for arrays.

ndrwrbgs reviewed (changes requested):
From what I see cursorily, it looks like there's a lot more changes than the title suggestions. Perhaps you could send a more focused/isolated review so that each of the changes can be evaluated on it's own merit (rather than one change having to block any benefit from the others due to being coupled)?

If my understanding of part of the change is correct about 'generic function' actually being 'use IReadOnlyList' then I'd be concerned about performance regressions when Array, Span, and List are passed in, as those have compiler/JIT optimizations for iteration.
